### DISCUSSION TITLE
Monitoring and Debugging in Static Web Apps

### DISCUSSION THREAD

**Discussion by Reshmi-Sriram** [Azure/static-web-apps#851] [Ideas]
Hey folks!

We are working on improving the monitoring and debugging aspects of Static Web Apps. We'd love to hear from you on:

- How do you like the current support?
- What are the things you'd like to see to improve your debugging experience?
- What are the key monitoring variables that concerns your application usage?

Below is the proposed schema for monitoring, namely `StaticSiteHTTPLogs` and `StaticSiteDiagnosticLogs`. 
Let us know if the below schema solves your common asks!

## StaticSiteHTTPLogs: Schema
```js
{
  "time": "2022-02-20T02:20:22Z",
  "resourceId": "/SUBSCRIPTIONS/SUBSCRIPTIONID/RESOURCEGROUPS/RESOURCEGROUPNAME/PROVIDERS/MICROSOFT.WEB/STATICSITES/SITENAME",
  "location": "global",
  "operationName": "Microsoft.Web/StaticSites/Read",
  "category": "StaticSiteHttpLogs",
  "resultType": "Failed", // Succeeded for 2xx and 3xxx, Failed for 4xx and 5xxx 
  "resultSignature": "404", // HTTP Status Code
  "resultDescription": "Page not found",
  "durationMs": 100, // time taken to serve the request in milliseconds
  "callerIpAddress": "192.168.1.1",
  "correlationId": "336a0049-d98f-48ca-8b21-d39244c34486", //ID used to correlate logs for a given request
  "uri": "https://yellow-coast-05f303010.1.azurestaticapps.net/api/hello?input=world", // Request Uri
  "properties": {
    "httpMethod": "GET", // HTTP Method. can be GET, HEAD, OPTIONS.
    "kiloBytesSent": 10,
    "kiloBytesReceived": 520,
    "userAgent": "Mozilla/5.0 (platform; rv:geckoversion) Gecko/geckotrail Firefox/firefoxversion",
    "referrer": "https://example.com/",
    "userDetails": {
      "rolesCount": 3,
      "identityProvider": "github",
    },
    "requestHandler": {
      "type": "Backend", // Can be StaticFile, Backend, Authentication. Note: AFD won't come till here for logging
      "properties": {
        "statusCode": "500",
        "errorMessage": "Error message from backend"
      }
    }
  }
}
```

## StaticSiteDiagnosticLogs: Schema
```js
{
  "time": "2022-02-20T02:20:22Z",
  "resourceId": "/SUBSCRIPTIONS/SUBSCRIPTIONID/RESOURCEGROUPS/RESOURCEGROUPNAME/PROVIDERS/MICROSOFT.WEB/STATICSITES/SITENAME",
  "location": "global", // Mandatory field
  "operationName": "Microsoft.Web/StaticSites/Deployment", // can also AssignFunction, Authentication etc.
  "category": "StaticSiteDiagnosticLogs",
  "correlationId": "336a0049-d98f-48ca-8b21-d39244c34486", //ID used to correlate logs for a given request
  "level": "Informational", // The severity level of the event. Must be one of Informational, Warning, Error, or Critical.
  "properties": {
     "message": "Use this to add any message for this log"
   }
}
```

**Comment by philip-coats-softbank** [comment]
at a mininum, i would just want a way to get a page hit log (timestamp, url, and user id) so we can monitor usage.

**Comment by Danish-Mentorange** [comment]
The key is AppInsights. If we are building a SaaS application using Static Web App then it should definitely log the user & session data in the AppInsights (which as per my research currently doesn't do that).
Any frontend lives on the AppInsights stats such as DAU, MAU, user flow, traces, etc. and without that information we feel handicapped.

So requesting you to please log more in the AppInsights that is connected to the SWA. Thanks.

**Comment by andreascorti** [comment]
Hello @Reshmi-Sriram, 
I would like to know if there are any plans to add Diagnostic Settings to monitor the devOps of the applications, especially including the logs today are presented only in the portal "Diagnose and Solve Problems". 
Our security teams would like to set alerts if an application receives many 4xx errors, and also would like to have a better history of the deployments done and configuration changes.

Thanks

**Comment by o-l-a-v** [comment]
The documentation is lacking when it comes to diagnostic logs and SWA.

* <https://learn.microsoft.com/en-us/azure/azure-monitor/reference/supported-logs/microsoft-web-staticsites-logs>

Compared to Recovery Services Vault:

* <https://learn.microsoft.com/en-us/azure/azure-monitor/reference/supported-logs/microsoft-recoveryservices-vaults-logs>

It does not say what Log Analytics table(s) are supported, and there are noe example queries.

I've tried to enable `StaticSiteDiagnosticLogs` and `StaticSiteHttpLogs` to a Log Analytics Workspace like so, sensitive info redacted.

```json
{
  "EventHubAuthorizationRuleId": null,
  "EventHubName": null,
  "Id": "/subscriptions/<redacted>/resourcegroups/<redacted>/providers/microsoft.web/staticsites/<redacted>/providers/microsoft.insights/diagnosticSettings/<redacted>",
  "Log": [
    {
      "Category": "StaticSiteDiagnosticLogs",
      "CategoryGroup": null,
      "Enabled": true,
      "RetentionPolicyDay": 0,
      "RetentionPolicyEnabled": false
    },
    {
      "Category": "StaticSiteHttpLogs",
      "CategoryGroup": null,
      "Enabled": true,
      "RetentionPolicyDay": 0,
      "RetentionPolicyEnabled": false
    }
  ],
  "LogAnalyticsDestinationType": null,
  "MarketplacePartnerId": null,
  "Metric": [
    {
      "Category": "AllMetrics",
      "Enabled": false,
      "RetentionPolicyDay": 0,
      "RetentionPolicyEnabled": false,
      "TimeGrain": null
    }
  ],
  "Name": "<redacted>",
  "ServiceBusRuleId": null,
  "StorageAccountId": null,
  "SystemDataCreatedAt": null,
  "SystemDataCreatedBy": null,
  "SystemDataCreatedByType": null,
  "SystemDataLastModifiedAt": null,
  "SystemDataLastModifiedBy": null,
  "SystemDataLastModifiedByType": null,
  "Type": "Microsoft.Insights/diagnosticSettings",
  "WorkspaceId": "/subscriptions/<redacted>/resourceGroups/<redacted>/providers/Microsoft.OperationalInsights/workspaces/<redacted>"
}
```

But I can't find any logs yet. This was hours ago. Do we have to use dedicated table? 🤔
